lgc/state/PalMetadata.cpp

```diff
@@ -0,0 +1,253 @@
+/*
+ ***********************************************************************************************************************
+ *
+ *  Copyright (c) 2020 Advanced Micro Devices, Inc. All Rights Reserved.
+ *
+ *  Permission is hereby granted, free of charge, to any person obtaining a copy
+ *  of this software and associated documentation files (the "Software"), to deal
+ *  in the Software without restriction, including without limitation the rights
+ *  to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+ *  copies of the Software, and to permit persons to whom the Software is
+ *  furnished to do so, subject to the following conditions:
+ *
+ *  The above copyright notice and this permission notice shall be included in all
+ *  copies or substantial portions of the Software.
+ *
+ *  THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ *  IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ *  FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ *  AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ *  LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ *  OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+ *  SOFTWARE.
+ *
+ **********************************************************************************************************************/
+/**
+ ***********************************************************************************************************************
+ * @file  PalMetadata.cpp
+ * @brief LLPC source file: PalMetadata class for manipulating PAL metadata
+ ***********************************************************************************************************************
+ */
+#include "lgc/state/PalMetadata.h"
+#include "lgc/state/PipelineState.h"
+#include "lgc/state/TargetInfo.h"
+#include "llvm/IR/Metadata.h"
+#include "llvm/IR/Module.h"
+
+#define DEBUG_TYPE "lgc-pal-metadata"
+
+using namespace lgc;
+using namespace llvm;
+
+namespace {
+
+// The name of the metadata node containing PAL metadata. This name is part of the interface from LGC into
+// the LLVM AMDGPU back-end when compiling for PAL ABI.
+const char PalMetadataName[] = "amdgpu.pal.metadata.msgpack";
+
+// PAL metadata SPI register numbers for the start of user data.
+// First the ones that only apply up to GFX8.
+constexpr unsigned int mmSPI_SHADER_USER_DATA_LS_0 = 0x2D4C;
+// Up to GFX9 only
+constexpr unsigned int mmSPI_SHADER_USER_DATA_ES_0 = 0x2CCC;
+// Then the ones that apply to all hardware.
```

Comment:
This is a nitpick, but I'm not sure I agree with this phrasing. If you're going to say that gfx9 puts the GS data into the ES registers, you *should* also be saying that it puts the HS data into the LS registers.

That doesn't necessarily reduce the total amount of confusion, but it aligns the confusion with what's in the register definitions according to the hardware.

Reply:
Right, maybe I've confused this in a different way. I need to go back and check, but I think what I found is that the mmSPI_SHADER_USER_DATA_HS_0 definition I have here works on all HW, with gfx9+ putting LS+HS into the HS regs. The confusing bit is that the gfx9 documentation actually calls those regs LS, even though they have the same numbers as what other arches call HS. So I ignored the gfx9 documentation. :-)

Perhaps I should explicitly point that out.